### PR TITLE
refactor: use icon-button component for close icons FE-2508

### PR DESCRIPTION
**Pull request by damianOfKyan**
### Proposed behaviour
Introduces new component 'icon-button' (`<IconButton/>`).
Fixes rendering issues for close icon across multiple components, fix is around close icon on focus behaviour specifically border display. Thanks to the new component that it brings one source of truth across multiple components that use it.

Screenshot 2020-01-20 at 17 39 00

Promoting composition over inheritance.

```
<IconButton
    onAction={ () => {} }
>
    <Icon type='close' />
</IconButton>
```

IconButton component will accept only one component as its `children` prop and its `<Icon>`.

### Current behaviour
Close icon was re-styled for every component that was consuming Icon component. This was causing us to use/implement repetitive code. There were number of render bugs across components related to close icon and focus border. The border on focus would either not display at all, or would display incorrectly. 

Screenshot 2020-01-20 at 17 45 35

### Checklist
- [x] Release notes (Conventional Commits) 
- [x] Unit tests
- [ ] Cypress automation tests
- [x] Storybook added or updated
 - [x] Theme support
- [x] Fix Alert component (close-icon border on focus)
- [x] Fix Dialog-Full-Screen component (close-icon border on focus)
- [x] Fix Dialog component (close-icon border on focus)
- [x] Fix Message component (close-icon border on focus)
- [x] Fix Pages component (close-icon border on focus)
- [x] Fix Sidebar component (close-icon border on focus)
- [x] Fix Toast component (close-icon border on focus)
- [x] Theme support (close-icon border on focus)

### CODE REVIEW

File: src/components/icon/icon-close.spec.js [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+      assertStyleMatch(
+        {
+          color: baseTheme.icon.disabled
+        },
+        wrapper.toJSON()
+      );
+    });
+
+    describe.each(['error', 'info', 'business'])(
+      'whent the background theme is %s',
```
Comment (DipperTheDan) [inline src/components/icon/icon-close.spec.js]: Spelling mistake? Should be 'when'

Comment (damianOfKyan) [reply src/components/icon/icon-close.spec.js]: fixed in cd6c59f

File: src/components/icon/icon-close.spec.js [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+            verticalAlign: 'middle'
+          },
+          wrapper.toJSON(),
+          { modifier: '&::before' }
+        );
+      });
+    }
+  );
+
+  describe('when the icon type is "sevices', () => {
```
Comment (ianforster-sage) [inline src/components/icon/icon-close.spec.js]:
```suggestion
  describe('when the icon type is "services', () => {
```

Comment (damianOfKyan) [reply src/components/icon/icon-close.spec.js]: Fixed in 9076bd0

File: src/components/icon/icon-close.spec.js [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+        assertStyleMatch(
+          {
+            color: baseTheme.colors.primary
+          },
+          wrapper.toJSON()
+        );
+
+        assertStyleMatch(
+          {
+            color: '#1E861E'
```
Comment (grabkowski) [inline src/components/icon/icon-close.spec.js]: we should test colors with theme objects instead of the direct output. A change in theme will lead to test failure.

File: src/components/icon/icon-close.spec.js [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+    describe.each(['info', 'error', 'success', 'warning'])(
+      'when bgTheme is set to one of the statuses',
+      (status) => {
+        const wrapper = renderStyles({ bgTheme: status });
+        const hoverColors = {
+          info: '#005C9B',
+          error: '#9F2C3F',
+          success: '#008C00',
+          warning: '#BA5000'
```
Comment (grabkowski) [inline src/components/icon/icon-close.spec.js]: hardcoded colors

File: src/components/icon/icon-close.style.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      display: block;
+    }
+
+    &:focus {
+      ${!isClassic(theme) && css`
+        outline: solid 3px ${theme.colors.focus};
+      `}
+      ${isClassic(theme) && css`
+        outline: -webkit-focus-ring-color auto 5px;
+      `}
```
Comment (grabkowski) [inline src/components/icon/icon-close.style.js]: could you override that style in `classicIconStyles`?

File: src/components/message/message.style.js [hunk 6]
```diff
@@ -18,6 +21,37 @@ const MessageStyle = styled.div`
...
+      ${({ theme }) => css`
+        &:focus {
+          color: ${theme.text.color};
+          background-color: transparent;
+          ${!isClassic(theme) && css`
+            outline: solid 3px ${theme.colors.focus};
+          `}
+          ${isClassic(theme) && css`
+            outline: -webkit-focus-ring-color auto 5px;
+          `}
```
Comment (grabkowski) [inline src/components/message/message.style.js]: you could just override that style

Comment (damianOfKyan) [reply src/components/message/message.style.js]: resolved in b526110

File: src/components/icon/icon-close.style.js (hunk 5 above)
Comment (damianOfKyan) [reply src/components/icon/icon-close.style.js]: resolved in b526110

File: src/components/icon/icon-close.spec.js (hunk 4 above)
Comment (damianOfKyan) [reply src/components/icon/icon-close.spec.js]: resolved in b526110

File: src/components/icon/icon-close.spec.js (hunk 3 above)
Comment (damianOfKyan) [reply src/components/icon/icon-close.spec.js]: resolved in b526110

File: src/components/icon/icon-close.spec.js [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+        type: 'services', theme: classicTheme, fontSize: 'small'
+      });
+      assertStyleMatch(
+        {
+          marginTop: '-4px'
+        },
+        wrapper.toJSON(),
+        { modifier: '&::before' }
+      );
+    });
```
[the remark this answers is not in the document]
Comment (damianOfKyan) [reply src/components/icon/icon-close.spec.js]: I updated two of these to be in descrive.each, as 3rd one had different margin assert. resolved in b526110

File: src/components/icon/icon-close.spec.js [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+      assertStyleMatch(
+        {
+          marginTop: '-7px'
+        },
+        wrapper.toJSON(),
+        { modifier: '&::before' }
+      );
+    });
+
+    it('it applies additional margin-top styling when the browser is safari and fontSize is "small"', () => {
```
[the remark this answers is not in the document]
Comment (damianOfKyan) [reply src/components/icon/icon-close.spec.js]: resolved in b526110

File: src/components/icon/icon-close.component.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        bgSize={ this.props.bgSize }
+        bgShape={ this.props.bgShape }
+        bgTheme={ this.props.bgTheme }
+        fontSize={ this.props.fontSize }
+        disabled={ this.props.disabled }
+        type={ this.type }
+        key='icon'
+        className={ this.props.className || null }
+        { ...this.componentProps }
+        { ...tagComponent('icon', this.props) }
```
[the remark this answers is not in the document]
Comment (damianOfKyan) [reply src/components/icon/icon-close.component.js]: resolved in b526110

File: src/components/dialog-full-screen/icon.style.js [hunk 10]
```diff
@@ -12,15 +13,26 @@ const StyledIcon = styled(Icon)`
...
     color: ${({ theme }) => theme.icon.onLightBackgroundHover};
   }
 
+  &:focus {
+    ${({ theme }) => !isClassic(theme) && css`
+      outline: solid 3px ${theme.colors.focus};
+    `}
+    ${({ theme }) => isClassic(theme) && css`
+      outline: -webkit-focus-ring-color auto 5px;
+    `}
```
[the remark this answers is not in the document]
Comment (damianOfKyan) [reply src/components/dialog-full-screen/icon.style.js]: resolved in b526110

**Comment by harpalsingh** [comment]
@damianOfKyan I looked at the dialog component for this, but this isn't applied to the fullscreen dialog?

**Comment by mkrds** [comment]
I might be out of the loop on this one, but could you tell me what is the reason for having two basically same components `Icon` and `CloseIcon` where the only difference is that `CloseIcon` has a `type = 'close'` set up by default?

File: src/components/dialog-full-screen/dialog-full-screen.spec.js [hunk 11]
```diff
@@ -160,6 +159,51 @@ describe('DialogFullScreen', () => {
...
+        );
+        const foundIconClose = wrapper.find(IconClose).first();
+        expect(foundIconClose.exists).toBeTruthy();
+        expect(foundIconClose.find(Icon).first().hasClass('dialog-close-icon')).toBeTruthy();
+      });
+    });
+  });
+
+  describe('when showCloseIcon is false', () => {
+    it('doeos not render close icon', () => {
```
Comment (DipperTheDan) [inline src/components/dialog-full-screen/dialog-full-screen.spec.js]: Spelling mistake. It should be "does".

Comment (damianOfKyan) [reply src/components/dialog-full-screen/dialog-full-screen.spec.js]: fixed in 5665da7

File: src/components/sidebar/sidebar.spec.js [hunk 12]
```diff
@@ -62,12 +65,58 @@ describe('Sidebar', () => {
...
+            />
+          );
+          expect(wrapper.find(Icon).exists).toBeTruthy();
+          expect(wrapper.find(Icon).first().hasClass('sidebar-close-icon')).toBeTruthy();
+        });
+      });
+    });
+
+    describe('when showCloseIcon is false', () => {
+      it('doeos not render close icon', () => {
```
Comment (DipperTheDan) [inline src/components/sidebar/sidebar.spec.js]: Spelling mistake. Should be "does".

Comment (damianOfKyan) [reply src/components/sidebar/sidebar.spec.js]: fixed in 6471717

File: src/components/icon/icon-close.spec.js [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+          foundIconClose.simulate('keyDown', keyDownParams);
+          expect(onCancel).toHaveBeenCalledTimes(0);
+        });
+      });
+    });
+
+    describe('IconClose onClick event', () => {
+      it('calls close callback', () => {
+        const foundIconClose = wrapper.find(Icon).first();
+        // const keyDownParams = { keyCode: 9, which: 9 };
```
Comment (ianforster-sage) [inline src/components/icon/icon-close.spec.js]: please remove commented lines of code

Comment (damianOfKyan) [reply src/components/icon/icon-close.spec.js]: fixed 8c9a675

File: src/components/dialog-full-screen/content-classic.style.js [hunk 14]
```diff
@@ -12,7 +12,7 @@ export default ({ theme, hasHeader }) => isClassic(theme) && css`
 
   ${!hasHeader && `
     padding-top: 0;
-    margin-top: 0;
+    margin-top: 0px;
```
Comment (grabkowski) [inline src/components/dialog-full-screen/content-classic.style.js]: is that change needed?

File: src/components/icon/icon-close.spec.js [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+          onConfirm={ onConfirm }
+          open
+          showCloseIcon
+          subtitle='Test'
+          title='Test'
+          ariaRole='dialog'
+          closeIcon={ (
+            <Icon
+              type='close'
+              dataElement='close'
```
Comment (grabkowski) [inline src/components/icon/icon-close.spec.js]: `data-element`

File: src/components/icon/icon-close.spec.js [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+      );
+    });
+
+    describe('when closeIcon is provided with className', () => {
+      it('renders with className', () => {
+        const className = 'abc';
+        const icon = (
+          <Icon
+            type='close'
+            dataElement='close'
```
Comment (grabkowski) [inline src/components/icon/icon-close.spec.js]: `data-element`

File: src/components/alert/alert.stories.js [hunk 17]
```diff
@@ -97,6 +98,13 @@ function makeButtonStory(name, themeSelector) {
           stickyFormFooter={ stickyFormFooter }
           subtitle={ subtitle }
           open={ store.get('open') }
+          closeIcon={ (
+            <Icon
+              type='close'
+              dataElement='close'
```
Comment (grabkowski) [inline src/components/alert/alert.stories.js]: that prop should be in kebab-case: `data-element`

File: src/components/dialog-full-screen/dialog-full-screen.spec.js [hunk 18]
```diff
@@ -160,6 +159,51 @@ describe('DialogFullScreen', () => {
...
+            open
+            onCancel={ () => {} }
+            onConfirm={ () => {} }
+            title='Test'
+            data-role='baz'
+            data-element='bar'
+            closeIcon={ (
+              <Icon
+                type='close'
+                dataElement='close'
```
Comment (grabkowski) [inline src/components/dialog-full-screen/dialog-full-screen.spec.js]: `data-element`

File: src/components/dialog-full-screen/dialog-full-screen.stories.js [hunk 19]
```diff
@@ -55,6 +57,14 @@ function makeStory(name, themeSelector) {
             disableEscKey={ disableEscKey }
             ariaRole={ ariaRole }
             onClick={ handleClick }
+            showCloseIcon={ showCloseIcon }
+            closeIcon={ (
+              <Icon
+                type='close'
+                dataElement='close'
```
Comment (grabkowski) [inline src/components/dialog-full-screen/dialog-full-screen.stories.js]: `data-element`

File: src/components/dialog-full-screen/dialog-full-screen.stories.js [hunk 20]
```diff
@@ -93,6 +104,14 @@ function makeStickyFooterStory(name, themeSelector) {
...
```
Comment (grabkowski) [inline src/components/dialog-full-screen/dialog-full-screen.stories.js]: `data-element`

File: src/components/dialog/dialog.component.js [hunk 21]
```diff
@@ -161,19 +161,27 @@ class Dialog extends Modal {
...
-          onBlur={ this.onCloseIconBlur }
-        />
+    if (!this.props.showCloseIcon) return null;
+    const { closeIcon, onCancel } = this.props;
+
+    if (!closeIcon) {
+      const extraProps = {
+        onBlur: this.onCloseIconBlur,
+        type: 'close',
+        dataElement: 'close',
```
Comment (grabkowski) [inline src/components/dialog/dialog.component.js]: `data-element`

File: src/components/dialog/dialog.spec.js [hunk 22]
```diff
@@ -488,6 +491,43 @@ describe('Dialog', () => {
...
+              subtitle='Test'
+              size='small'
+              className='foo'
+              onCancel={ onCancel }
+              onConfirm={ () => {} }
+              showCloseIcon
+              closeIcon={ (
+                <Icon
+                  type='close'
+                  dataElement='close'
```
Comment (grabkowski) [inline src/components/dialog/dialog.spec.js]: `data-element`

File: src/components/dialog/dialog.stories.js [hunk 23]
```diff
@@ -71,6 +72,13 @@ function makeStory(name, themeSelector) {
             disableEscKey={ disableEscKey }
             ariaRole={ ariaRole }
             onClick={ handleClick }
+            closeIcon={ (
+              <Icon
+                type='close'
+                dataElement='close'
```
Comment (grabkowski) [inline src/components/dialog/dialog.stories.js]: `data-element`

File: src/components/flash/flash-legacy.component.js [hunk 24]
```diff
@@ -233,7 +233,7 @@ class Flash extends React.Component {
         </FlashMessageStyle>
         {!this.props.timeout && (
           <FlashCloseStyle
-            data-element='close'
+            dataElement='close'
```
Comment (grabkowski) [inline src/components/flash/flash-legacy.component.js]: `data-element`

File: src/components/icon/icon-close.component.js [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+  }, [onKeyDownProp]);
+
+  const onClick = useCallback((e) => {
+    onClickProp(e);
+  });
+
+  const extraProps = {
+    ...rest,
+    ref,
+    className: (className && classNames(className, 'close-icon')) || 'close-icon',
```
Comment (grabkowski) [inline src/components/icon/icon-close.component.js]: do we still need this `close-icon` class? The component should be styled with `styled-components`.

File: src/components/button-icon/button-icon.component.js [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+      e.stopPropagation();
+    }
+  };
+
+  const buttonRef = React.createRef();
+
+  const handleOnAction = (e) => {
+    e.preventDefault();
+    onAction(e);
+  };
```
Comment (grabkowski) [inline src/components/button-icon/button-icon.component.js]: is there a reason for preventing default on click?

File: src/components/button-icon/button-icon.component.js [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+
+  return (
+    <StyledButtonIcon
+      { ...rest }
+      onKeyDown={ onKeyDown }
+      onClick={ handleOnAction }
+    >
+      <Icon
+        type={ type }
+        ref={ buttonRef }
```
Comment (grabkowski) [inline src/components/button-icon/button-icon.component.js]: could you tell me why do we need this ref?

File: src/components/button-icon/button-icon.component.js [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+  );
+};
+
+ButtonIcon.defaultProps = {
+  type: 'close'
+};
+
+ButtonIcon.propTypes = {
+  /** Icon type */
+  type: PropTypes.string,
```
Comment (grabkowski) [inline src/components/button-icon/button-icon.component.js]: I think that the type should be required and `close` should be not set as default

File: src/components/alert/alert.spec.js [hunk 29]
```diff
@@ -1,37 +1,32 @@
...
     wrapper = mount(
-      <Alert
-        open
-        onCancel={ onCancel }
-        title='Alert title'
-        subtitle='Alert Subtitle'
-        data-element='bar'
-        data-role='baz'
-      />
+      <ThemeProvider theme={ baseTheme }>
```
[the remark this answers is not in the document]
Comment (damianOfKyan) [reply src/components/alert/alert.spec.js]: noted, thanks! fixed in: 28fd4d6

File: src/components/button-icon/button-icon.component.js (hunk 26 above)
Comment (damianOfKyan) [reply src/components/button-icon/button-icon.component.js]: not as far as I can see, added on autopilot without thinking it through, thanks!
fixed in: ea86715

File: src/components/button-icon/button-icon.component.js (hunk 27 above)
Comment (damianOfKyan) [reply src/components/button-icon/button-icon.component.js]: it is redundant, I left it there by accident, apologies.
fixed in: b996854

File: src/components/button-icon/button-icon.component.js (hunk 28 above)
Comment (damianOfKyan) [reply src/components/button-icon/button-icon.component.js]: agree! Type added in: 55e0442
updated respective components to pass in `type='close'`
a844a07
e1581ba
f4eca6b
df4b990
650287d

**Comment by ianforster-sage** [comment]
This commit https://github.com/Sage/carbon/pull/2608/commits/77a7dd029d9331525805919ffc3e86488d1a69a6 is now included in PR https://github.com/Sage/carbon/pull/2647
I'd say it should be removed from this PR and this whole PR is blocked until `2647` is merged.

File: src/components/sidebar/sidebar.component.js [hunk 30]
```diff
@@ -17,33 +17,22 @@ class Sidebar extends Modal {
...
+  closeIcon = () => {
+    const { showCloseIcon, onCancel } = this.props;
+    if (!showCloseIcon || !onCancel) return null;
+    return (
+      <IconButton
+        onAction={ onCancel }
+      >
+        <Icon
+          type='close'
+          className='carbon-sidebar__close-icon'
```
Comment (jamime) [inline src/components/sidebar/sidebar.component.js]: This PR removes some class names but has left this one in. We should be consistent with our approach.

File: src/components/pages/pages.stories.js [hunk 31]
```diff
@@ -111,6 +111,7 @@ function makeStory(name, themeSelector) {
           <DialogFullScreen
             open={ store.get('open') }
             onCancel={ handleCancel }
+            showCloseIcon
```
Comment (jamime) [inline src/components/pages/pages.stories.js]: Adding this would indicate that we have made a breaking change somewhere.

Comment (damianOfKyan) [reply src/components/pages/pages.stories.js]: I am discarding this commit. that entire commit I think is left over by accident from early version where I did breaking changes and I wanted to add knob to pages for and few other components to control if close icon should be shown or not.

**Comment by damianOfKyan** [comment]
This PR has been split into individual PR for each component affected:

https://github.com/Sage/carbon/pull/2690
https://github.com/Sage/carbon/pull/2691
https://github.com/Sage/carbon/pull/2692
https://github.com/Sage/carbon/pull/2693
https://github.com/Sage/carbon/pull/2694
